File: webauthn-server-core/src/main/scala/com/yubico/webauthn/impl/X5cAttestationStatementVerifier.scala
package com.yubico.webauthn.impl
import java.security.cert.X509Certificate

import com.fasterxml.jackson.databind.node.ArrayNode
import com.yubico.u2f.data.messages.key.util.CertificateParser
import com.yubico.webauthn.data.AttestationObject


trait X5cAttestationStatementVerifier {

  protected def getX5cAttestationCertificate(attestationObject: AttestationObject): Option[X509Certificate] =
    getAttestationTrustPath(attestationObject) flatMap { _.headOption }

  def getAttestationTrustPath(attestationObject: AttestationObject): Option[List[X509Certificate]] =
    attestationObject.attestationStatement.get("x5c") match {
      case certs: ArrayNode =>
        Some((for {
          i <- 0 until certs.size
          binary = certs.get(i)
        } yield {
          if (binary.isBinary)
            CertificateParser.parseDer(certs.get(i).binaryValue)
          else
            throw new IllegalArgumentException(
              s"""Each element of "x5c" property of attestation statement must be a binary value, was: ${binary.getNodeType}"""
            )
        }).toList)

      case _ => None
    }

}

File: webauthn-server-core/src/main/java/com/yubico/webauthn/impl/X5cAttestationStatementVerifier.java
package com.yubico.webauthn.impl;

import com.fasterxml.jackson.databind.JsonNode;
import com.yubico.u2f.data.messages.key.util.CertificateParser;
import com.yubico.webauthn.data.AttestationObject;
import java.io.IOException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;


public interface X5cAttestationStatementVerifier {

    default Optional<X509Certificate> getX5cAttestationCertificate(AttestationObject attestationObject) throws IOException, CertificateException {
        return getAttestationTrustPath(attestationObject).flatMap(certs -> certs.stream().findFirst());
    }

    default Optional<List<X509Certificate>> getAttestationTrustPath(AttestationObject attestationObject) throws IOException, CertificateException {
        JsonNode x5cNode = attestationObject.getAttestationStatement().get("x5c");

        if (x5cNode != null && x5cNode.isArray()) {
            List<X509Certificate> certs = new ArrayList<>(x5cNode.size());

            for (JsonNode binary : x5cNode) {
                if (binary.isBinary()) {
                    certs.add(CertificateParser.parseDer(binary.binaryValue()));
                } else {
                    throw new IllegalArgumentException(String.format(
                        "Each element of \"x5c\" property of attestation statement must be a binary value, was: %s",
                        binary.getNodeType()
                    ));
                }
            }

            return Optional.of(certs);
        } else {
            return Optional.empty();
        }
    }

}
